### DISCUSSION TITLE
Type annotations for common element props

### DISCUSSION THREAD

**Discussion by numpde** [idom-team/idom#917] [Ideas]
I'd like to propose a class to facilitate a little bit the writing of [component props](https://beta.reactjs.org/reference/react-dom/components/common#common-props), especially for those coming from ReactJS or HTML/JS. It should be basically a dict but with the benefit of code completion. Something like this:

```Python
# https://beta.reactjs.org/reference/react-dom/components/common#common
@dataclasses.dataclass
class Props(dict):
    NULL = object()  # analogous to JS 'null'
    UNDEFINED = None  # analogous to JS 'undefined'
    MISSING = object()  # means "not present at all"
    EMPTY = object()  # no value given (as in <button disabled />), but this is probably redundant

    # Special React props
    children: list = MISSING  # children -- ?
    dangerously_set_inner_html: dict = MISSING  # dangerouslySetInnerHTML
    ref: Any = MISSING  # ref
    suppress_content_editable_warning: bool = MISSING  # suppressContentEditableWarning
    suppress_hydration_warning: bool = MISSING  # suppressHydrationWarning
    style: dict = MISSING  # style

    # Standard DOM props
    access_key: str = MISSING  # accessKey
    aria: dict = MISSING  # aria-* (expanded)
    aria_: str = MISSING  # aria-* (any aria- attribute)
    auto_capitalize: str = MISSING  # autoCapitalize
    class_name: str = MISSING  # className
    content_editable: bool = MISSING  # contentEditable
    # etc.
```

Note, this would also simplify conditional prop supply. Maybe it could specialize to [input](https://beta.reactjs.org/reference/react-dom/components/input#props) and such using `Props[html.input]`. The `children` prop is debatable.

Furthermore (but this is not specific to this suggestion/class), 'data' and 'aria' props could optionally be dictionaries somewhat similar to 'style', and 'className' could accept a list, e.g.,
```Python
Props(
    data={'bs': {'toggle': "collapse"}},
    class_name=["btn", "btn-primary"],
)
``` 
would become
```Python
{data_bs_toggle="collapse", class_name="btn btn-primary"}
```
before being processed further.

With the pre-1.0.0 syntax, it'd work as follows:
```Python
html.div(
    Props(
        class_name="container",
    ),
    ...  # children
)
````
and with the 1.0.0a3 syntax:
```Python
html.div(
    ...  # children,
    **Props(
        class_name="container",
    )
)
````

**Comment by rmorshea** [comment]
> 'data' and 'aria' props could optionally be dictionaries

I'm not 100% sure we should do it just because it breaks from what React does, but I quite like this idea.

**Reply by Archmonger** [reply]
Can we accomplish autocomplete with a TypedDict type hint?

**Comment by rmorshea** [comment]
> ```python
> @dataclasses.dataclass
> class Props(dict):

Instead of using dataclasses, we'd probably use a `TypedDict` instead. But yes, I think this something @Archmonger was talking about doing.

**Reply by numpde** [reply]
I observe (in Python 3.11.1) that 

```Python
from typing import TypedDict

class X(TypedDict):
    a: int = 1  # IDE complains that RHS not supported in TypedDict
    b: int

    def __new__(cls, **kw):
        return dict(**kw)

# IDE complains that z is unexpected, but the assert passes (even without __new__)
assert X(b=2, z=0) == {'b': 2, 'z': 0}  
```

which I think is suboptimal.

**Reply by Archmonger** [reply]
The goal for TypeDict to autocomplete the keys while typing. It can't be used to set any default values.

It's used as a type hint for a regular dict, and isn't something that the user will manually construct through a class-like interface.

See this PEP for more details: https://peps.python.org/pep-0589/

**Reply by rmorshea** [reply]
![Peek 2023-02-09 12-04](https://user-images.githubusercontent.com/4041990/217924953-f112eafe-7e07-4224-847f-1d154eacafc2.gif)

**Reply by rmorshea** [reply]
In VSCode I have to manually request the type hint with Ctrl+Space, but there's a discussion to change that [here](https://github.com/microsoft/pylance-release/discussions/3710) (needs votes to actually be implemented though).

**Reply by numpde** [reply]
Probably better [like this](https://peps.python.org/pep-0589/#alternative-syntax) if you want to allow reserved keywords (and easier to auto-generate): 
```Python
Movie = TypedDict('Movie', {'name': str, 'year': int}, total=False)
```

I guess, `dict[str, Any]` is to allow arbitrary props, as well.

PyCharm (recent) also hints only on Ctrl+Space.

**Reply by Archmonger** [reply]
I mentioned in one of my posts that good TypeDict support in editors is still hit or miss.

I hope for that to change soon.

**Reply by rmorshea** [reply]
You can specify `total=False` via the class version as well:

```python
class Movie(TypedDict, total=False): ...
```

**Reply by Archmonger** [reply]
https://gist.github.com/Archmonger/8d01d2efba1071f4950549004719eb19

I've created a Python script that parses/converts the [React-DOM props from the docs](https://beta.reactjs.org/reference/react-dom/components/common#common-props)  into a `snake_case` `TypeDict` via some heuristics.

Props that IDOM does not support (such as `children`, `ref`) are automatically parsed out.

Output looks something like this

```python
{
 'access_key': <class 'str'>,
 'aria-': None,
 'auto_capitalize': <class 'str'>,
 'class_name': <class 'str'>,
 'content_editable': <class 'bool'>,
 'dangerously_set_inner_html': dict[str, int | str | float | list | dict | bool | None],
}
```

**Reply by rmorshea** [reply]
This looks great! Small correction, event handlers should be typed as `Callable[..., Awaitable[None]] | Callable[..., None]`.

**Reply by Archmonger** [reply]
Fixed!

On a subjective note, I left `data-` and `aria-` autocomplete suggestions in (as seen in the example above). However, the literal `"aria-"` is technically not a valid prop.

I haven't really decided if they should stay or not. Let me know your opinion.

The React docs mention them, so that makes me feel like they should stay. But on the flip side, it's not really a valid prop unless something is proceeding (`aria-my-value`).

**Reply by Archmonger** [reply]
I'm going to try to figure out a way to incorporate [all CSS attributes](https://www.w3schools.com/cssref/index.php) and [all basic HTML attributes](https://www.w3.org/TR/REC-html40/index/attributes.html) as well

**Reply by rmorshea** [reply]
That would be pretty awesome to have.

**Reply by Archmonger** [reply]
https://gist.github.com/Archmonger/8d01d2efba1071f4950549004719eb19

I've added CSS style parsing to the gist. Unfortunately, I wasn't able to find a convenient data source to do type inference here, so everything is `str | int`.

The `TypeDict` assumes that since we are avoiding `camelCase` for react props, we will be doing the same for style dicts.

**Reply by numpde** [reply]
One option could be to reverse-engineer a validator like [this one](https://github.com/w3c/css-validator/tree/main/org/w3c/css/properties/css3).

**Reply by Archmonger** [reply]
The amount of LOCs in there is a bit overkill.

I'd rather go line-by-line on CSS properties and manually add types to them than reverse engineer a java CSS validator.

For now, I think the generic types are alright.

**Reply by Archmonger** [reply]
https://gist.github.com/Archmonger/8d01d2efba1071f4950549004719eb19

Updated the script to parse the HTML5 attribute spec with type heuristics.

```python
{'a': {'download': <class 'str'>,
       'href': <class 'str'>,
       'hreflang': <class 'str'>,
       'ping': <class 'str'>,
       'referrerpolicy': <class 'str'>,
       'rel': <class 'str'>,
       'target': <class 'str'>,
       'type': <class 'str'>},
 'abbr': {'title': <class 'str'>},
 'button': {'disabled': <class 'bool'>,
            'form': str | int,
            'formaction': <class 'str'>,
            'formenctype': <class 'str'>,
            'formmethod': <class 'str'>,
            'formnovalidate': <class 'bool'>,
            'formtarget': <class 'str'>,
            'name': <class 'str'>,
            'popoverhidetarget': str | int,
            'popovershowtarget': str | int,
            'popovertoggletarget': str | int,
            'type': <class 'str'>,
            'value': <class 'str'>},
  ...
}

**Reply by rmorshea** [reply]
This will be a bit tricky. There are attributes like `autocomplete` that, in React, are written as `autoComplete` instead. Also, all `-` separated style attributes should be `_` instead.

**Reply by Archmonger** [reply]
I think we can avoid those clashes during merging the two dicts. When we merge, we should ensure that `key.strip("_")` doesn't exist as a key in either dict.

Or we can manually exclude known clashes.

**Reply by rmorshea** [reply]
To clarify, when I say `autocomplete` is written as `autoComplete` in React, that means that on the Python side, we'll want users to declare it as `auto_complete` (in order for our camelCase conversion to work). This isn't really about conflicts so much as it is about finding all cases where we need to add a `_` delimiter to HTML attributes.
